homeassistant/components/husqvarna_automower_ble/config_flow.py

```diff
@@ -60,62 +64,218 @@ async def async_step_bluetooth(
+
+        return self.async_create_entry(
+            title=title,
+            data={
+                CONF_ADDRESS: self.address,
+                CONF_CLIENT_ID: channel_id,
+                CONF_PIN: self.pin,
+            },
+        )
+
+    async def async_step_reauth(
+        self, entry_data: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Perform reauthentication upon an API authentication error."""
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
         self, user_input: dict[str, Any] | None = None
     ) -> ConfigFlowResult:
-        """Handle the initial step."""
-        if user_input is not None:
+        """Confirm reauthentication dialog."""
+        errors: dict[str, str] = {}
+
+        if user_input:
             self.address = user_input[CONF_ADDRESS]
-            await self.async_set_unique_id(self.address, raise_on_progress=False)
-            self._abort_if_unique_id_configured()
-            return await self.async_step_confirm()
+            self.pin = user_input[CONF_PIN]
+
+            try:
+                device = bluetooth.async_ble_device_from_address(
+                    self.hass, self.address, connectable=True
+                ) or await get_device(self.address)
+
+                (channel_id, mower) = await self.connect_mower(device)
+
+                response_result = await mower.connect(device)
+                await mower.disconnect()
+                if (
+                    response_result is ResponseResult.INVALID_PIN
+                    or response_result is ResponseResult.NOT_ALLOWED
+                ):
+                    errors["base"] = "invalid_auth"
+                elif response_result is not ResponseResult.OK:
+                    errors["base"] = "cannot_connect"
+                else:
+                    data = {
+                        CONF_ADDRESS: self.address,
+                        CONF_CLIENT_ID: channel_id,
+                        CONF_PIN: self.pin,
+                    }
+
+                    return self.async_update_reload_and_abort(
+                        self._get_reauth_entry(), data=data
+                    )
+
+            except (TimeoutError, BleakError):
+                return self.async_abort(reason="cannot_connect")
+
+        user_input = {}
+
+        if self.address:
+            user_input[CONF_ADDRESS] = self.address
+        if self.pin:
+            user_input[CONF_PIN] = self.pin
 
         return self.async_show_form(
-            step_id="user",
-            data_schema=vol.Schema(
-                {
-                    vol.Required(CONF_ADDRESS): str,
-                },
+            step_id="reauth_confirm",
+            data_schema=self.add_suggested_values_to_schema(
+                vol.Schema(
+                    {
+                        vol.Required(CONF_ADDRESS): str,
+                        vol.Required(CONF_PIN): int,
```

Comment:
Can you confirm that this is really an `int`? This way I cannot input PINs with leading zeros like `0000` and `0011`. Also I think the length should be limited. On other integrations `PIN` is used with `str`.

Reply:
The PIN is sent as a number to the mower, not as a string. I think the input in the flow should be a string though for the reason you mention, the config flow should then try to convert it to an int and fail if that's not possible.

Reply:
I modified the config flow to revert back to requesting a string from the user, but this time there's validation for non-numerical PIN.